### ISSUE TITLE
Server Side Blazor support fallback to other protocols

### ISSUE THREAD

**Issue by spaolo1969** [aspnet/AspNetCore#6920]
It would be very nice to have Server Side Blazor implement fallback to other SignalR protocols (SSE, lp...) when a websockets connection cannot be established.

This is actually almost always the case when deploying in Enterprise environment, websockets connection are very seldom established because of the proxies and firewalls implemented.

**Comment by chrdlx** [comment]
Does this mean that if your IIS Server or browser doesn't support WebSockets then Razor Components won't work??   
 I just downloaded the latest Preview 3, and made a Razor Components solution... after publishing it to my local IIS the interactivity in the Counter page doesn't work at all.  It seems it only works with websocket enabled.
 If this is the case then it's very bad news for me,  since I wanted to deploy a small razor components app to my company's server which is a 2008 R2 which as far as I know,  it won't support never WebSockets =( =( =( =(
 
Also I remember (almost sure) that it worked fine in previous versions using longpolling, did this break in the latest preview 3?

Thank you

**Comment by mkArtakMSFT** [comment]
@anurse isn't SignalR connection automatically handle this?

**Comment by BrennanConroy** [comment]
They might be hitting other problems https://github.com/aspnet/AspNetCore/issues/8384

**Comment by anurse** [comment]
> @anurse doesn't SignalR connection automatically handle this?

Yes, SignalR should be doing this automatically. If we fail to connect to the WebSocket endpoint from the client it should automatically fall back through the other transports. I don't know how/if Razor Components has configured SignalR to know for sure if something is interfering with that. It's also possible it **is** falling back and some other issue is occurring (as @BrennanConroy) referenced.

Getting more logs from the client might help show us what is happening (i.e. if SignalR is trying to fallback or if something is going wrong in that logic). We just posted a new article on [Logging and diagnostics in SignalR](https://docs.microsoft.com/en-us/aspnet/core/signalr/diagnostics?view=aspnetcore-2.2) with information on how to collect more detailed logs that may help identify issues (again, this applies to SignalR, not sure how these config options can be flowed through to Razor Components since they create the HubConnectionBuilder.

@mkArtakMSFT it might help to have a doc describing how to collect this kind of diagnostic information in Razor Components (possibly referencing the SignalR one I linked and just expanding on the Razor Components stuff) since these issues can be quite common and generally require a little more logging/diagnostic information to diagnose.

**Comment by spaolo1969** [comment]
Actually my issue is quite old (Jan 22) and it might be that other previews worked with fallback, didnt have time to check. This morning i took some time to check. IIS with websockets working as expected. if i uninstall websocket protocol from IIS the UI is not even shown in the browser .

Looking at the network tab you can see it does negotiate protocols

{connectionId: "GRne4_Dm5hYP4eMBpo0j7w",…}
availableTransports: [{transport: "ServerSentEvents", transferFormats: ["Text"]},…]
connectionId: "GRne4_Dm5hYP4eMBpo0j7w"

But it doesnt work. 

Last autumn I was eager to see how it worked and i had to install Telerik Fiddler to understand what a "websockets tunnel" is. So if it uses websockets tunnel it'd work only with websockets i thought. Having some experience in deploying signalr apps in enterprise scenarios i know websockets can be impossibile to obtain (not that they dont install it on IIS, problems are from proxies, firewalls and so on, i am no networking expert but network folks arent good listeners, lets put it like that :) )

**Comment by mkArtakMSFT** [comment]
Thanks @anurse. We plan to address exposing SignalR hub configuration as part of https://github.com/aspnet/AspNetCore/issues/8277

Seems like without a way to configure those options we won't be able to move forward here. So marking this investigation as blocked until we figure out how we expose the configuration options for the Hub.

@rynowak if you have any suggestion here which can unblock this investigation earlier, please share.

**Comment by anurse** [comment]
One note is that the issue you linked is for configuring hub options **on the server**. In this case you'd need to provide a way to configure the **JavaScript client** connection options to produce more logging.

**Comment by mkArtakMSFT** [comment]
I see. @pranavkm, @javiercn do we have a way for configuring SinglarR Javascript client?

**Comment by pranavkm** [comment]
Not as yet. Although we could cover it as part of - https://github.com/aspnet/AspNetCore/issues/6887

**Comment by danroth27** [comment]
This should be fixed now by https://github.com/aspnet/AspNetCore/issues/8679.
